### ISSUE TITLE
hyperparameter optimization

### ISSUE THREAD

**Issue by stevie1023** [stevie1023/cspn-gpspn#6]
Hi Martin, Merry Christmas and Happy new year! Sorry to bother you during the holiday and sorry for the delay of updates these days..jjust reply when you feel free~ I tried to do global optimization as you said last time but the results are somehow not satisfying.  Take parkinsons dataset where the output y is two-dimensional so I ended up with two GP models for all the leaves. I tried to aggregate the gradients of all leaves (in my case I took the log marginal likelihoods) and propagate that to the root in every step. To be more specific, I took pure sum of mlls for product nodes and split nodes, weighted sum for sum nodes, and then do mll_root backward and optimizer.step. The best result is as follows: 
![image](https://user-images.githubusercontent.com/51037813/103170852-889ef800-4882-11eb-85a5-d8e09f82daba.png)
The leaf size for above is 2000, the number of children for sum nodes is 2. 
If I set the threshold for leaf size smaller or more children for sum nodes, the results will be worse by around 0.07 for this dataset.
I am wondering why the results barely improve but degrade?  Should I take sum of mlls for product nodes or where can I possibly mess up in the optimization? I Literally have no idea = = SOS
 Enjoy your vocation~

**Comment by trappmartin** [comment]
Hi, sorry for the late reply. I didn't see your message.

Could you point me to the lines in the code where you do the hyper parameter optimisation and the computations?

**Comment by stevie1023** [comment]
Sure~ Here is what I do for global optimization.

https://github.com/stevie1023/cspn-gpspn/blob/5d3461b9f8a5ea9a60c526fb6efd7394afd9dd54/prod_spngp.py#L67

发件人: Martin Trapp<mailto:notifications@github.com>
发送时间: 2021年1月18日 16:57
收件人: stevie1023/cspn-gpspn<mailto:cspn-gpspn@noreply.github.com>
抄送: Stevie Zhu<mailto:zzmmyy1023@outlook.com>; Author<mailto:author@noreply.github.com>
主题: Re: [stevie1023/cspn-gpspn] hyperparameter optimization (#6)


Hi, sorry for the late reply. I didn't see your message.

Could you point me to the lines in the code where you do the hyper parameter optimisation and the computations?

―

**Comment by trappmartin** [comment]
So, the `update_mll` procedure is calculating bottom up the final log marginal likelihood in your code. Is this correct?

If so, you need to perform a log sum exp operation at the sum nodes. I'm not entirely sure what you do know in: https://github.com/stevie1023/cspn-gpspn/blob/5d3461b9f8a5ea9a60c526fb6efd7394afd9dd54/prod_inference.py#L55-L59

What does the `@` do?

Also, it is probably best to learn a separate set of hyper parameters for each Y_i. So if you have a 2D output and want to train the length scale parameters you would obtain one length scale parameter for Y_1 and one for Y_2.
